src/ql/arch/cc/pass/gen/vq1asm/detail/codegen.cc

```diff
@@ -544,250 +724,432 @@ void Codegen::customGate(
             }
-            if (breg_operands.size() > 1) {
-                QL_FATAL(
-                    "Readout instruction '" << qasm(iname, operands, breg_operands)
-                    << "' requires 0 or 1 bit operands, not " << breg_operands.size()
+            if (ops.bregs.size() > 1) {
+                QL_INPUT_ERROR(
+                    "Readout instruction requires 0 or 1 bit operands, not " << ops.bregs.size()
                 );
             }
-
-            // store operands
-            if (settings.getReadoutMode(iname)=="feedback") {
-                bi.isMeasFeedback = true;
-                bi.operands = operands;
-                //bi.creg_operands = creg_operands;    // NB: will be empty because of checks performed earlier
-                bi.breg_operands = breg_operands;
+#endif
+            // flag this bundle as performing real-time measurements, and store operands
+            bi.isMeasRsltRealTime = true;
+
+            // handle classic operand
+            if (ops.bregs.empty()) {    // FIXME: currently always
+                bi.breg_operand = ops.qubits[0];                    // implicit classic bit for qubit
+                QL_IOUT("using implicit bit " << bi.breg_operand << " for qubit " << ops.qubits[0]);
+            } else {    // FIXME: currently impossible
+                bi.breg_operand = ops.bregs[0];
+                QL_IOUT("using explicit bit " << bi.breg_operand << " for qubit " << ops.qubits[0]);
             }
         }
 
-        // store 'expression' for conditional gates
-        bi.condition = condition;
-        bi.cond_operands = cond_operands;
-#endif
+        // Handle the condition. NB: the 'condition' field exists for all conditional_instruction sub types,
+        // but we only handle it for custom_instruction
+        tInstructionCondition instrCond = decode_condition(operandContext, custom.condition);
+        bi.instructionCondition = instrCond;
 
-        QL_DOUT("customGate(): iname='" << iname <<
+        QL_DOUT("custom_instruction(): iname='" << iname <<
              "', duration=" << durationInCycles <<
              " [cycles], instrIdx=" << csv.si.instrIdx <<
              ", group=" << csv.si.group);
 
-        // NB: code is generated in bundleFinish()
+        // NB: code is generated in bundle_finish()
     }   // for(signal)
+}
 
-#if OPT_PRAGMA
-    RawPtr<const Json> pragma = settings.getPragma(iname);
-    if (pragma) {
-        for (Vec<BundleInfo> &vbi : bundleInfo) {
-            // FIXME: for now we just store the JSON of the pragma statement in bundleInfo[*][0]
-            if(vbi[0].pragma) {
-                QL_FATAL("Bundle contains more than one gate with 'pragma' key");    // FIXME: provide context
-            }
-            vbi[0].pragma = pragma;
+/************************************************************************\
+| Structured control flow
```

Comment:
OpenQL supports a 'structure decomposer' pass to decompose structured control flow to blocks and gotos.
The idea behind that is to have reusable code for this that need not be specialized in the codegen. What is the reason that this pass was not used?

Reply:
I think there are 2 main reasons: 

- because the original program structure is lost, debugging becomes difficult
- specific optimisations become difficult. The Q1 processor has a 'loop' instruction that decrements a register and loops if the result is non-zero. The CC backend now uses that instruction when possible, doing so would be difficult if structure has been decomposed.

Reply:
This issue is not urgent, merely to reduce later development time, e.g. when one of the two structure decomposition methods is to be synchronized with the other one because of certain benefits. So, I'm exploring now the reasons behind so that we learn from this.

Keeping track on correspondence between input and output program, can be done through code location annotations.

Yes, the structural decomposition pass is all or nothing. In the case of CC with static loop support, one would want lowering of if-then-else, repeat-until, while-do, dynamic-for but not static loop support. This could be established by adding options to the structural decomposition pass to selectively enable the respective individual lowerings. Would you consider this an option?

Reducing code duplication saves debugging time and future development time (you know of course).
For instance dealing with the following: nested static loops, updates to the IR, adding if-convert (turning cheap if-then-elses to conditionally executed code).

So, your opinion please.

Reply:
Options to the structural decomposition pass may be of help, although I imagine it may be cleaner to rewrite the decomposition pass according to the lines of https://tree-gen.readthedocs.io/en/latest/interpreter.gen.html#visitor-pattern, and then override from the decomposition class in the backend wherever the default isn't suitable.

Then again the 'problem' is far worse for expressions than it is for structured control flow.

https://github.com/QuTech-Delft/OpenQL/issues/424 contains quite a few good ideas to handle commonalities in a better fashion

Reply:
Separating walking down the tree and visiting the nodes, allowing overriding the latter, is a better way to do this, indeed.
We need to make a real step forward here.
Next to the visitor pattern, we need a transform pattern that works similarly but returns a replacement IR fragment at each level.
And in both we need a context parameter (kind of stack) to do context dependent visits/transformations.
Furthermore, the IR may contain structures that are not trees; the related classes should also be extended with visitor-like methods; those then maintain the sequencing through the non-tree. Similarly, for trees there are several walks possible rather then the depth-first, left-to-right one. I didn't look but perhaps the visitor pattern can be instantiated for each of these. E.g. each node maybe visited twice, before recursing and after recursing down its children.
This what concerns the visitor pattern.

Code generation for statements and expression trees is indeed far from trivial. Look at all those 'code generators' around. Please indicate when you think you need to take a step back and add more support for this. For one thing, I learned that it is easier to make multiple traversals, gradually annotating and transforming a tree rather than doing everything in one go. It may be illustrative when I once give a high-level overview of code generation in CoSy. CoSy has passes, visitor/transform patterns, generated pattern matcher to transform to target code based on configuration, storage allocation passes, pass plug-ins, etc.  CoSy lets itself be read in a top-down form so perfect for learning.

Reply:
In parallel to your work, we are rewriting old-IR passes to new-IR passes, e.g. clifford and later mapper.
In those contexts I will try to make this step forward as well.

Reply:
I think it would be good to keep an open eye on commonalities when more passes are migrated to the new IR. The CC was the first to make the switch I guess, and it leans heavily on new_to_old. Since that is to be removed in the end, I didn't particular care about some copy/paste/edit steps that I wouldn't normally find acceptable.